src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs

```diff
@@ -520,14 +520,40 @@ private static bool IsAllDigits(string filterText)
                 CompletionItem bestFilterMatch,
                 SnapshotPoint caretPosition,
                 CompletionHelper completionHelper,
-                CompletionFilterReason reason)
+                CompletionFilterReason filterReason)
             {
-                if (bestFilterMatch == null || model.UseSuggestionMode)
+                var itemViewSpan = model.GetViewBufferSpan(bestFilterMatch.Span);
+                var fullFilterText = model.GetCurrentTextInSnapshot(itemViewSpan, caretPosition.Snapshot, endPoint: null);
+                var textSpan = itemViewSpan.TextSpan;
+
+                // Switch to soft selection, if user moved caret to the start of a non-empty filter span.
+                // This prevents commiting if user types a commit character at this position later, but 
+                // still has the list if user types filter character
+                // i.e. blah| -> |blah -> !|blah
+                // We want the filter span non-empty because we still want hard selection in the following case:
+                //
+                //  A a = new |
+                if (caretPosition == textSpan.Start && textSpan.Length > 0)
                 {
                     return false;
                 }
 
-                var textSnapshot = caretPosition.Snapshot;
+                return IsHardSelection(fullFilterText, model.Trigger.Kind, bestFilterMatch, completionHelper, filterReason, this.Controller.GetRecentItems(), model.UseSuggestionMode);
+            }
+
+            internal static bool IsHardSelection(
+                string fullFilterText,
+                CompletionTriggerKind initialTriggerKind,
+                CompletionItem bestFilterMatch,
+                CompletionHelper completionHelper,
+                CompletionFilterReason filterReason,
+                ImmutableArray<string> recentItems,
+                bool useSuggestionMode)
+            {
+                if (bestFilterMatch == null || useSuggestionMode)
+                {
+                    return false;
+                }
```

Comment:
why the changes in this file?

Reply:
I reused some methods from the file. To make them accessible from the new code, I had to make those methods: a) internal instead of private and b) static

---
In reply to: [232139265](https://github.com/dotnet/roslyn/pull/29016#discussion_r232139265) [](ancestors = 232139265)